flight/Modules/Logging/logging.c

```diff
@@ -381,25 +353,135 @@ static void loggingTask(void *parameters)
 				read_sector = loggingData.FileSectorNum;
 			}
 			LoggingStatsSet(&loggingData);
-
+			// fall-through to default case
+		default:
+			//  Makes sure that we are not hogging the processor
+			PIOS_Thread_Sleep(10);
+			// Close the file if necessary
+			if (write_open) {
+				PIOS_STREAMFS_Close(streamfs_id);
+				loggingData.MinFileId = PIOS_STREAMFS_MinFileId(streamfs_id);
+				loggingData.MaxFileId = PIOS_STREAMFS_MaxFileId(streamfs_id);
+				LoggingStatsSet(&loggingData);
+				write_open = false;
+			}
 		}
+	}
+}
 
-		i++;
+/**
+ * Forward data from UAVTalk out the serial port
+ * \param[in] data Data buffer to send
+ * \param[in] length Length of buffer
+ * \return -1 on failure
+ * \return number of bytes transmitted on success
+ */
+static int32_t send_data(uint8_t *data, int32_t length)
+{
+	if( PIOS_COM_SendBuffer(logging_com_id, data, length) < 0)
+		return -1;
+
+	written_bytes += length;
+
+	return length;
+}
+
+
+/**
+ * Function for sorting linked list based on logging period and makes
+ * sure that objects with private queues come first.
+ */
+static int info_cmp(struct UAVOLogInfo *info_1, struct UAVOLogInfo *info_2)
+{
+	if ((info_1->queue == shared_queue) && (info_2->queue != shared_queue)){
+		return true;
+	}
+	if ((info_1->queue != shared_queue) && (info_2->queue == shared_queue)){
+		return false;
 	}
+
+	return info_1->logging_period > info_2->logging_period;
 }
 
+/**
+ * @brief Callback for adding an object to the logging queue
+ * @param ev the event
+ */
+static void obj_updated_callback(UAVObjEvent * ev)
+{
+	struct UAVOLogInfo *info;
+	uint32_t now;
+
+	if (loggingData.Operation != LOGGINGSTATS_OPERATION_LOGGING){
+		// We are not logging, so all events are discarded
+		return;
+	}
+	LL_FOREACH(log_info, info) {
```

Comment:
So things are sorted by logging interval... but this is unfortunate because it has nothing to do with the interval at which our callback will get called.  Unless I'm misreading this, this has paradoxical effects-- I could lower the interval at which gyros are logged, which bumps it to the end of the list, which in turn could greatly increase CPU utilization.

I think the interval/next time to log probably belong at the callback layer.  The benefits there are that there's no need to search an additional list at all and it's possible to entirely avoid these issues and remove the need for the list.  The downside is that it increases the memory use for all other callbacks even if logging is not used.  There's ways around this though-- for example the event mask could be updated to have a bit of HAS_MAX_INTERVAL where we allocate an additional trailer structure containing the last time and the interval.